File: Min Stack.java
/**
 * Solution to https://leetcode.com/problems/min-stack/
 */
import java.util.Stack;

class MinStack {
  Stack<Integer> stack = new Stack<Integer>();
  Stack<Integer> min_list = new Stack<Integer>();
  public void push(int x) {        
    if(min_list.empty() || x<=min_list.peek())            
      min_list.push(x);
    stack.push(x);
  }
  public void pop() {
    if(!stack.empty() && stack.pop().equals(min_list.peek()))
      min_list.pop();
  }
  public int top() {
    return stack.peek();
  }
  public int getMin() {
    return min_list.peek();
  }
}

File: Min Stack.js
/**
 * Solution to https://github.com/ericpony/leetcode
 */
/**
 * @constructor
 */
var MinStack = function() {
    this.stack = [];
    this.min = [];
};

Array.prototype.peek = function() {
    return this[this.length-1];
};

/**
 * @param {number} x
 * @returns {void}
 */
MinStack.prototype.push = function(x) {
    this.stack.push(x);
    if(this.min.length)
        this.min.push(Math.min(this.min.peek(), x));
    else
        this.min.push(x);
};

/**
 * @returns {void}
 */
MinStack.prototype.pop = function() {
    this.min.pop();
    return this.stack.pop();
};

/**
 * @returns {number}
 */
MinStack.prototype.top = function() {
    return this.stack.peek();
};

/**
 * @returns {number}
 */
MinStack.prototype.getMin = function() {
    return this.min.peek();
};
